Доработана функциональность по техническому заданию 9 спринта

asdbl opened the pull request.

asdbl committed:
Добавить доступ к методам программы через HTTP API.

TyutterinYakov commented on src/server/BaseHttpHandler.java:
```diff
@@ -0,0 +1,88 @@
+
+    public void sendWrongMethod(HttpExchange exchange) throws IOException {
+        exchange.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        exchange.sendResponseHeaders(400, 0);
```
Для отсутствующего метода запроса принято возвращать статус 405:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405

TyutterinYakov commented on src/server/EpicHandler.java:
```diff
@@ -0,0 +1,99 @@
+        String method = exchange.getRequestMethod();
+        String[] pathSplit = exchange.getRequestURI().getPath().split("/");
+        switch (method) {
+            case "GET":
```
Для методов можно было бы ввести отдельный enum, чтобы не работать со строковыми константами)

TyutterinYakov commented on src/server/EpicHandler.java:
```diff
@@ -0,0 +1,99 @@
+
+    public void deleteEpic(HttpExchange exchange, int id) {
+        try {
+            if (manager.getEpics().contains(manager.getTaskById(id))) {
```
Действительно ли требуется использовать метод `contains` для полученного списка?)
Почему бы не проверить то, что вернул метод `getTaskById`. Аналогично в других местах)

TyutterinYakov commented on src/server/SubtaskHandler.java:
```diff
@@ -0,0 +1,99 @@
+    public void getSubtaskByIdResponse(HttpExchange exchange, int id) {
+        try {
+
+            if (manager.getTaskById(id) != null) {
```
Можно было бы использовать более краткую запись:
```
final Subtask subtask (Subtask) manager.getTaskById(id);
if (subtask == null) {
    sendText(exchange, "Подзадача не найдена!", 404);
    return;
}
sendText(exchange, gson.toJson(subtask), 200);
```

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -170,7 +169,7 @@ private boolean isDuplicate(Epic epic) {
     public void update(Task task) {
         if (isOverlap(task)) {
-            task.setStartTime(getPrioritizedTasks().getLast().getEndTime());
+            throw new RuntimeException("Пересечение задачи по времени!");
```
Лучше использовать для этой цели более конкретное исключение, можно даже ввести свое) Так как `RuntimeException` является слишком высокоуровневым

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -190,18 +189,17 @@ public void update(Epic epic) {
     public void update(Subtask subtask) {
         if (isOverlap(subtask)) {
-            subtask.setStartTime(getPrioritizedTasks().getLast().getEndTime());
+            throw new RuntimeException("Пересечение времени задачи");
```
Лучше использовать для этой цели более конкретное исключение, можно даже ввести свое) Так как `RuntimeException` является слишком высокоуровневым

TyutterinYakov commented on src/server/EpicHandler.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) {
```
(предложение) Кажется, что эту логику можно было бы не дублировать и объявить этот метод в родительском классе, где мы будем проверять лишь метод запроса:
```
switch (method) {
     case "GET": ...
```

На каждый кейс будет в родителе отдельный метод, по дефолту каждый из которых будет возвращать статус 405 METHOD NOT ALLOWED - https://timeweb.com/ru/community/articles/oshibka-405-method-not-allowed-chto-za-oshibka-i-kak-ee-ispravit
А затем, уже в наследниках, те методы, что используются - будут переопределены и в них будет реализована необходимая логика) Таким образом дублирование кода снизится)

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

asdbl committed:
Добавить enum для методов Http.
Добавить исключение пересечения времени TaskOverlapException.
Исправить дублирование кода в Handler.

TyutterinYakov commented on src/exceptions/TaskOverlapException.java:
Лучше придерживаться единого стиля именования пакетов) Либо все в ед. числе, либо же во мн.)

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей финального спринта этого проекта) У тебя получилась классная реализация, молодец! Желаю сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного дня! Да пребудет с тобой Spring!🎇🌅
